File: internal/service/attribution/claimcode.go
package attribution

import (
	"crypto/rand"
	"math/big"
	"strings"
)

// claimAlphabet 领奖码字符集。
//
// 刻意排除 0/O/1/I/l —— iOS 上用户需要手动输入领奖码，而那是 iOS 侧唯一可计费
// 的归因路径。一个字符的误读就是一次收入损失，所以宁可牺牲码空间。
// 32 个字符 × 8 位 ≈ 1.1e12 组合，配合限流足够抗枚举。
const claimAlphabet = "23456789ABCDEFGHJKMNPQRSTUVWXYZ"

const claimCodeLen = 8

// NewClaimCode 生成一个领奖码。使用 CSPRNG —— 领奖码直接对应可计费转化，
// 可预测的码等于可伪造的收入。
func NewClaimCode() (string, error) {
	return randomString(claimAlphabet, claimCodeLen)
}

// trackingAlphabet 投放位 ID 的字符集。
const trackingAlphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

const trackingIDLen = 10

// NewTrackingID 生成投放位 ID。
//
// 必须不可枚举：如果 KOL 能猜出别人的链接，就能互相刷量或窃取归因。
// 这里不做人类可读优化 —— tracking_id 从不需要手动输入。
func NewTrackingID() (string, error) {
	return randomString(trackingAlphabet, trackingIDLen)
}

// NormalizeClaimCode 把用户输入的领奖码规整为存储形态：大小写不敏感、
// 去除用户或输入法可能带入的分隔符与空白。
//
// 手动输入是 iOS 侧的必经路径，容错直接影响核销完成率（W7 的核心验证指标）。
// 这里刻意不做混淆字符映射：字符集本身已排除 0/O/1/I/L，所以这几个字符
// 出现在输入里必定是误输入，而 "0" 究竟想打 O 还是 D 无法可靠推断。
// 猜错会把一个有效码变成另一个有效码，归到错误的 KOL 名下 —— 宁可报错重输。
func NormalizeClaimCode(in string) string {
	var b strings.Builder
	b.Grow(len(in))
	for _, r := range strings.ToUpper(strings.TrimSpace(in)) {
		switch r {
		case ' ', '-', '_', '\t', '\n', '\r', '.':
			continue
		}
		b.WriteRune(r)
	}
	return b.String()
}

// ValidClaimCodeFormat 报告 code 是否符合领奖码的格式。
//
// 在查库前做格式校验，可以让绝大多数枚举尝试和输入错误在不消耗数据库资源的
// 情况下被拒绝，同时给用户一个比"码不存在"更有用的错误。
func ValidClaimCodeFormat(code string) bool {
	if len(code) != claimCodeLen {
		return false
	}
	for _, r := range code {
		if !strings.ContainsRune(claimAlphabet, r) {
			return false
		}
	}
	return true
}

func randomString(alphabet string, n int) (string, error) {
	max := big.NewInt(int64(len(alphabet)))
	b := make([]byte, n)
	for i := range b {
		idx, err := rand.Int(rand.Reader, max)
		if err != nil {
			return "", err
		}
		b[i] = alphabet[idx.Int64()]
	}
	return string(b), nil
}

File: apps/api/src/attribution/claim_code.rs
//! Claim-code and tracking-id generation, normalisation, and format validation.

use rand::Rng;

/// Claim-code alphabet.
///
/// Deliberately excludes `0 O 1 I L` — on iOS users type codes manually, the only billable path.
/// One misread character is lost revenue; smaller code space is acceptable.
/// 31 chars × 8 digits ≈ 8.5e11 combinations; rate limits handle enumeration.
const CLAIM_ALPHABET: &[u8] = b"23456789ABCDEFGHJKMNPQRSTUVWXYZ";
const CLAIM_CODE_LEN: usize = 8;

/// Tracking-id alphabet.
// TODO(console): remove these allows when placement issuance (KOL console) ships.
#[allow(dead_code)]
const TRACKING_ALPHABET: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
#[allow(dead_code)]
const TRACKING_ID_LEN: usize = 10;

/// Generate a claim code.
///
/// Cryptographically secure randomness — codes map to billable conversions; predictable codes forge revenue.
pub fn new_claim_code() -> String {
    random_string(CLAIM_ALPHABET, CLAIM_CODE_LEN)
}

/// Generate a tracking id.
///
/// Must be unguessable: if KOLs can guess each other's links, they can farm or steal attribution.
/// No human-readability optimisations — tracking_id is never typed manually.
#[allow(dead_code)]
pub fn new_tracking_id() -> String {
    random_string(TRACKING_ALPHABET, TRACKING_ID_LEN)
}

/// Normalise user input to storage form: case-insensitive, strip separators and whitespace users or
/// IMEs may insert.
///
/// Manual entry is mandatory on iOS; tolerance directly affects redemption completion (core W7 metric).
///
/// **No confusable-character mapping**: the alphabet already excludes `0 O 1 I L`, so these in input
/// are mistakes and `0` could mean `O` or `D` — unreliable. Guessing wrong maps one valid code to
/// another under the wrong KOL — better format error and retry.
pub fn normalize_claim_code(input: &str) -> String {
    input
        .trim()
        .chars()
        .filter(|c| !matches!(c, ' ' | '-' | '_' | '\t' | '\n' | '\r' | '.'))
        .flat_map(char::to_uppercase)
        .collect()
}

/// Whether a claim code matches the expected format.
///
/// Validate before hitting the database — rejects most enumeration and typos without DB cost and gives
/// a more useful error than "code not found".
pub fn is_valid_claim_code(code: &str) -> bool {
    code.len() == CLAIM_CODE_LEN && code.bytes().all(|b| CLAIM_ALPHABET.contains(&b))
}

fn random_string(alphabet: &[u8], len: usize) -> String {
    let mut rng = rand::thread_rng();
    (0..len)
        .map(|_| alphabet[rng.gen_range(0..alphabet.len())] as char)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    /// Claim alphabet must exclude confusable characters — manual iOS entry is the only billable path.
    #[test]
    fn claim_code_excludes_confusable_chars() {
        for _ in 0..500 {
            let code = new_claim_code();
            assert_eq!(code.len(), 8);
            for bad in ['0', 'O', '1', 'I', 'L'] {
                assert!(!code.contains(bad), "code {code} contains confusable {bad}");
            }
        }
    }

    #[test]
    fn claim_codes_do_not_repeat() {
        let codes: HashSet<String> = (0..200).map(|_| new_claim_code()).collect();
        assert_eq!(codes.len(), 200, "duplicate claim codes");
    }

    /// tracking_id must be unguessable — predictable links enable farming or attribution theft.
    #[test]
    fn tracking_ids_are_unpredictable() {
        let ids: HashSet<String> = (0..300).map(|_| new_tracking_id()).collect();
        assert_eq!(ids.len(), 300, "duplicate tracking_id");
        assert!(ids.iter().all(|id| id.len() == 10));
    }

    #[test]
    fn normalize_handles_common_input_noise() {
        for (input, want) in [
            ("ab3xy9zk", "AB3XY9ZK"),
            ("AB3X-Y9ZK", "AB3XY9ZK"),
            ("ab3x y9zk", "AB3XY9ZK"),
            (" AB3XY9ZK ", "AB3XY9ZK"),
            ("ab3x_y9zk", "AB3XY9ZK"),
        ] {
            assert_eq!(normalize_claim_code(input), want, "input {input:?}");
        }
    }

    /// No confusable mapping: guessing `0` as `O` or `D` can turn one valid code into another under the wrong KOL.
    #[test]
    fn normalize_does_not_guess_confusables() {
        let got = normalize_claim_code("AB0XY1ZK");

        assert_eq!(got, "AB0XY1ZK", "must not substitute 0 and 1");
        assert!(!is_valid_claim_code(&got), "should fail format validation");
    }

    #[test]
    fn format_validation() {
        assert!(is_valid_claim_code(&new_claim_code()));

        assert!(!is_valid_claim_code(""), "empty code");
        assert!(!is_valid_claim_code("AB3XY9Z"), "too short");
        assert!(!is_valid_claim_code("AB3XY9ZKQ"), "too long");
        assert!(!is_valid_claim_code("ab3xy9zk"), "lowercase not normalised");
        assert!(!is_valid_claim_code("00000000"), "out-of-alphabet chars");
    }
}
